### PR TITLE
MAINT: Ensure array_dealloc does not modify refcount of self

### PR DESCRIPTION
**Pull request by seberg**
This is even uglier then before in some ways, but avoids changing
the global totalrefcount in python debug builds.

---

Mostly for discussion. We probably do not want this, OTOH, not using INCREF is necessary if we want to allow third parties to debug nicely with `sys.gettotalrefcount` or `pytest-leaks` (at least if they use object arrays).

### CODE REVIEW

**Comment by seberg** [comment]
I pushed an alternative fix for discussion. We could just allocate the iterator object statically (basically using a borrowed reference). For the old style/simple iterator that is fairly simple.

**Comment by mattip** [comment]
I think this is fine. Since we should not be calling `PyArray_IterNew` inside a `tp_dealloc`, yet we still have to iterate over all the elements and decref them, it seems "stealing" a reference to a `PyArrayIterObject` this way is really the only way to do it.

**Comment by seberg** [comment]
Anyone wishes to chime in? I feel it is fine to create this internal helper (`PyArray_RawIterBaseInit`) to allow creating the old-style iterator as a local stack allocation, and – because it is guaranteed to be local – not need to hold on a reference to the array.

I am also somewhat fine with not doing it, if anyone feels it is a pattern we never use elsewhere and is just a bit too special to start using.

**Review by pv — COMMENTED** [review]
+1, fix itself looks necessary. I guess question about names is not important.

File: numpy/core/src/multiarray/iterators.c [hunk 1]
```diff
@@ -116,10 +116,12 @@ get_ptr_simple(PyArrayIterObject* iter, npy_intp *coordinates)
...
  * PyArrayNeighborhoodIterObject
  *
- * Increase ao refcount
+ * Steals a reference to the array object which gets removed at deallocation,
+ * if the iterator is allocated statically and its dealloc not called, it
+ * can be thought of as borrowing the reference.
  */
-static PyObject *
-array_iter_base_init(PyArrayIterObject *it, PyArrayObject *ao)
+void
```
[the remark this answers is not in the document]
Comment (seberg) [reply numpy/core/src/multiarray/iterators.c]: Yes, I think you are right. Changed that (and squashed the commits). I left the name for now, but I do not have much of an opinion. We sometimes use `PyArray_` for internal names, and sometimes not. I do not think we are very consistent about it.

[the remark this answers is not in the document]
Comment (pv) [reply numpy/core/src/multiarray/iterators.c]: Well, I think it's also ok as it is now.

**Comment by seberg** [comment]
@pv thanks, so we can just merge this?

**Comment by pv** [comment]
I think so. LGTM, and no smoke came out when using this.

**Comment by seberg** [comment]
OK, I will just merge it then. Thanks Pauli!

**Pull request merged**